File: android/Reguerta/app/src/main/java/com/reguerta/user/data/products/ChainedProductRepository.kt
package com.reguerta.user.data.products

import com.reguerta.user.domain.products.Product
import com.reguerta.user.domain.products.ProductRepository

class ChainedProductRepository(
    private val primary: ProductRepository,
    private val fallback: ProductRepository,
) : ProductRepository {
    override suspend fun getAllProducts(): List<Product> {
        val primaryProducts = primary.getAllProducts()
        return if (primaryProducts.isNotEmpty()) primaryProducts else fallback.getAllProducts()
    }

    override suspend fun getProductsForVendor(vendorId: String): List<Product> {
        val primaryProducts = primary.getProductsForVendor(vendorId)
        return if (primaryProducts.isNotEmpty()) primaryProducts else fallback.getProductsForVendor(vendorId)
    }

    override suspend fun upsertProduct(product: Product): Product {
        val fallbackUpdated = fallback.upsertProduct(product)
        val primaryUpdated = runCatching { primary.upsertProduct(product) }.getOrNull()
        return primaryUpdated ?: fallbackUpdated
    }
}

File: ios/Reguerta/Reguerta/Data/Products/ChainedProductRepository.swift
import Foundation

actor ChainedProductRepository<Primary: ProductRepository, Fallback: ProductRepository>: ProductRepository {
    private let primary: Primary
    private let fallback: Fallback

    init(primary: Primary, fallback: Fallback) {
        self.primary = primary
        self.fallback = fallback
    }

    func allProducts() async throws -> [Product] {
        let primaryProducts = try await primary.allProducts()
        if !primaryProducts.isEmpty {
            return primaryProducts
        }
        return try await fallback.allProducts()
    }

    func products(vendorId: String) async throws -> [Product] {
        let primaryProducts = try await primary.products(vendorId: vendorId)
        if !primaryProducts.isEmpty {
            return primaryProducts
        }
        return try await fallback.products(vendorId: vendorId)
    }

    func upsert(product: Product) async throws -> Product {
        try await primary.upsert(product: product)
    }
}
